numpy/core/src/multiarray/descriptor.c

```diff
@@ -1493,43 +1494,63 @@ _convert_from_any(PyObject *obj, int align)
         return ret;
     }
     else if (PyUnicode_Check(obj)) {
-        return _convert_from_str(obj, align);
+        if (Py_EnterRecursiveCall(
+            " while trying to create self-referential str dtypes" ) == 0) {
+            convert_ret = _convert_from_str(obj, align);
+        }
+        goto convert_done;
     }
     else if (PyTuple_Check(obj)) {
         /* or a tuple */
-        return _convert_from_tuple(obj, align);
+        if (Py_EnterRecursiveCall(
+            " while trying to create self-referential tuple dtypes" ) == 0) {
+            convert_ret = _convert_from_tuple(obj, align);
+        }
+        goto convert_done;
```

Comment:
This logic is wrong - it should be:
```suggestion
        if (Py_EnterRecursiveCall(
            " while trying to create self-referential tuple dtypes" ) != 0) {
            return NULL;
        }
        PyObject *ret = _convert_from_tuple(obj, align);
        Py_LeaveRecursiveCall();
        return ret;
```
`Py_LeaveRecursiveCall` must not be called if `Py_EnterRecursiveCall` fails.

Reply:
Oh sorry @eric-wieser , I'll read up a little more and apply the changes. Thanks a lot 👍